src/main/java/com/microsoft/sqlserver/jdbc/SQLServerBulkCopy.java

```diff
@@ -3163,6 +3197,24 @@ private boolean writeBatchData(TDSWriter tdsWriter) throws SQLServerException {
             // No more data available, return false so we do not execute any more batches.
             if (!goToNextRow())
                 return false;
+            
+            
+            if (insertRowByRow) {
+                // read response gotten from goToNextRow()
+                ((SQLServerResultSet) sourceResultSet).getTDSReader().readPacket();
+
+                // Create and send the initial command for bulk copy ("INSERT BULK ...").
+                tdsWriter = command.startRequest(TDS.PKT_QUERY);
+                String bulkCmd = createInsertBulkCommand(tdsWriter);
+                tdsWriter.writeString(bulkCmd);
```

Comment:
With cursor, we need to send data row by row, so it has to be inside the while loop. Also, to pass by MARS, the sending must happen after the reading, which is goToNextRow()